enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+removing the "initial" nodes of a cluster causes a lot of issues even in a
+regular, multi-node installation. So dealing with similar issues in single
+control-plane node clusters is also out of scope for this enhancement. For
+example, installing a cluster with 3 control-plane nodes and 3 workers results
+in unschedulable control-plane nodes. Scaling down the worker machine-sets from
+3 to 0 post-installation does not magically make the control-plane nodes
+schedulable, and as a result a lot of infrastructure workloads fail to schedule
+and operators degrade. User intervention is required to fix this.
+
+- Deal with expansion of the single-node control-plane by adding more
+control-plane nodes
+
+- Deal with expansion of clusters that have been installed before the
+implementation of this enhancement (if possible, their expansion may be
+addressed by an upgrade followed with documentation outlining the steps that
+the user has to take to enable expansion, e.g. patch the default
+`IngressController` to have node selector targetting the master pool rather
+than the worker pool).
+
+- Adjust the baremetal platform to support single control-plane node
+installations. The baremetal platform solves the "floating ingress" issue by
+using virtual IP addresses/keepalived.
+
+- Deal with users who want their ingress traffic to not go through the single
+control-plane node.
+
+- Deal with non-cloud, non-"none" platforms such as baremetal, vSphere, etc.
+
+## Proposal
+
+- Create a new `IngressPlacement` API parameter.
+
+- Make sure worker node ignition files are generated even in bootstrap-in-place
+single control-plane node "none"-platform installation.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I want to install a single control-plane
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+Introduce a new field in the Ingress config CR
+(`config.openshift.io/v1/ingresses`) called `IngressPlacement`.
+
+### Implementation Details/Notes/Constraints
+
+This new field will have one of these values - `ControlPlane` or `Workers`. 
+
+In addition, continue to allow the `.spec.replicas` and `.spec.nodePlacement`
+parameters in `operator.openshift.io/v1/ingresscontrollers` CRs to be omitted,
+but change the defaulting behavior for these fields.
+
+The value of the `IngressPlacement` field will affect the defaulting behavior
+of the `IngressController`'s `.spec.replicas` and `.spec.nodePlacement`
+parameters.  In the absence of an `IngressController` resource created by the
+user/installer, or when the user/installer creates an `IngressController` with
+these two parameters omitted, the Cluster Ingress Operator will choose the
```

Comment:
Do you mean that the values will be set on the object after create?

Have you considered a defaulting webhook for this rather than a controller? Using a controller implies that the change will be asynchronous. This means a user could create the object and observe their desired configuration without the defaults, then a moment later observe it with the results. Defaulting is better handled synchronously to avoid this discrepancy on the consumer side

Reply:
>Have you considered a defaulting webhook for this rather than a controller?

Since the router-default IngressController may be created by the Installer as an installer manifest, do "defaulting webhooks" this early in the cluster's lifecycle even exist / make sense?